added tag variable to mobkill

Osanchez opened the pull request:
I added a new dependency called NBTAPI, this allows devs to access the NBT information for not only items, but entities.
https://www.spigotmc.org/resources/nbt-api.7939/

The objective Entity Kill has been updated to accept a new parameter called tags. it is disabled by default. and defined like so
tags:my_custom_tag,my_custom_tag_2

When a mob kill objective (event) is fired and tags are defined, the objective will only complete if the mob that was killed contains all of the defined tags
---

### Related Issues
There are no related issues currently reported, but it fixes an issue I was having where a data pack boss I added was difficult to hook into an objective because I did not want to use mob name. 

### Requirements
- [X] I made sure my contribution fulfills the [requirements](https://docs.betonquest.org/DEV/Participate/Process/Submitting-Changes/#reviewers-checklist).

### Reviewer's checklist
Did the contributor...
- [ ]  ... test their changes?
- [x]  ... update the [Changelog](https://docs.betonquest.org/DEV/Participate/Process/Maintaining-the-Changelog/)?
- [ ]  ... update the [Documentation](https://docs.betonquest.org/DEV/Participate/Process/Docs/Workflow/)?
- [x]  ... adjust the [ConfigPatcher](https://docs.betonquest.org/DEV/API/ConfigPatcher)?
- [x]  ... solve all TODOs?
- [x]  ... remove any commented out code?
- [x]  ... add [debug messages](https://docs.betonquest.org/DEV/API/Logging/)?
- [x]  ... clean the commit history?

Check if the build pipeline succeeded for this PR!

Wolf2323 commented on src/main/java/org/betonquest/betonquest/objectives/MobKillObjective.java:
```diff
@@ -68,6 +75,15 @@ public void onMobKill(final MobKilledEvent event) {
         }
 
+        if (tags != null) {
+            final NBTEntity mcnbt = new NBTEntity(event.getEntity()); //Only for vanilla tags!
```
We do not want comments anymore at all in new code. I think this comment is not necessary at all.

Wolf2323 reviewed (changes requested):
In general this looks really good.
You missed to add Documentation for this.

And I only have one general point. As I have a partly similar thing i am currently developing.
This brings an alternative implementation in my mind.

What i am currently trying to do is implementing a `dataModify` event, similar to the minecraft data command. And i have two issues to solve. The first one is selecting the correct entities, the second one is selecting the entities.
For the first problem i found the `Bukkit.getServer().selectEntities()` method, that allows to use the minecraft selectors. This would one be one solution for you. But i am not sure if this is a good solution at all!
For the second issue i did not found an solution now, but may you newly added API could solve this.

May you have an opinion to this, I do not expect that you change you code now.

Wolf2323 commented on CHANGELOG.md:
```diff
@@ -127,6 +127,7 @@ to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
     - Version checks for ProtocolLib and Shopkeepers support
     - `mmoitemtake` event and `mmoitem` condition - now also check the backpack
         - this will not work until the item rework / until the backpack contains NBT data
+- entity kill Objective now supports `tags` as a parameter, which will need to match tags (NBT) of killed entity
```
Bring this line above
```
- Things that are also added in 1.12.X:
```

Osanchez commented:
> In general this looks really good. You missed to add Documentation for this.
> 
> And I only have one general point. As I have a partly similar thing i am currently developing. This brings an alternative implementation in my mind.
> 
> What i am currently trying to do is implementing a `dataModify` event, similar to the minecraft data command. And i have two issues to solve. The first one is selecting the correct entities, the second one is selecting the entities. For the first problem i found the `Bukkit.getServer().selectEntities()` method, that allows to use the minecraft selectors. This would one be one solution for you. But i am not sure if this is a good solution at all! For the second issue i did not found an solution now, but may you newly added API could solve this.
> 
> May you have an opinion to this, I do not expect that you change you code now.

removed comment, updated changelog, updated documentation.

Now to my recommendation for implementing features that involve NBT. I was hesitant to include a 3rd party API because interactions with NBT are dependent on MC version as you interact with NMS. It is nice that spigot provides some support for using vanilla selectors, but it also has very limited functionality (which you have experienced already). The API I provided has pretty good support and peaking into the API contents is also very careful when handling read/write with NBT because it incorporates file locks so multiple things cannot modify NBT data at the same time. These are edge cases that I find require a lot of experience working with NBT and the developer of this API has been involved in NBT for a very long time and is continuously updating this library. 

using this library you could follow similar syntax to mobkill

Wolf2323 commented on CHANGELOG.md:
```diff
@@ -127,6 +127,7 @@ to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
     - Version checks for ProtocolLib and Shopkeepers support
     - `mmoitemtake` event and `mmoitem` condition - now also check the backpack
         - this will not work until the item rework / until the backpack contains NBT data
+    - entity kill Objective now supports `tags` as a parameter, which will need to match tags (NBT) of killed entity
```
Sorry for the missunderstanding, but add this in line 121, directly above
> - Things that are also added in 1.12.X:

SaltyAimbOtter commented:
Hey @Osanchez, thank you for putting the time in to contribute.
Unfortunately I think that the downsides of this change heavily outweight the benefits.
Introducing a hard dependency on NBTAPI removes one of the core strategic aspects of BetonQuest: Version Independence. 

We have a zero-NMS policy which allows our community to instantly use BetonQuest when a new Minecraft version is released. By depending on NBTAPI we would have to wait for this API to be updated for newer versions.

Additionally, I'm unsure about the use case: Why do you need to check for NBT tags? As far as I know the [PersistentDataHolder](https://hub.spigotmc.org/javadocs/spigot/org/bukkit/persistence/PersistentDataHolder.html) API should be used to store custom data on any entity. Any vanilla entity property should be readable using the Spigot API.

Osanchez commented:
> Hey @Osanchez, thank you for putting the time in to contribute. Unfortunately I think that the downsides of this change heavily outweight the benefits. Introducing a hard dependency on NBTAPI removes one of the core strategic aspects of BetonQuest: Version Independence.
> 
> We have a zero-NMS policy which allows our community to instantly use BetonQuest when a new Minecraft version is released. By depending on NBTAPI we would have to wait for this API to be updated to support newer versions.
> 
> Additionally, I'm unsure about the use case: Why do you need to check for NBT tags? As far as I know the [PersistentDataHolder](https://hub.spigotmc.org/javadocs/spigot/org/bukkit/persistence/PersistentDataHolder.html) API should be used to store custom data on any entity. Any vanilla entity property should be readable using the Spigot API.

You already have dependencies on citizens and protocol lib which I believe already depends on NMS? 
PersistenTagContainer offers less, If I remember correctly this involves using scoreboard tags on entities. Again, I understand the concerns with version dependency. keep in mind that mythicmobs also uses NMS. its something that you could be made optional and enabled if the JAR is user provided.

SaltyAimbOtter commented:
The difference between these integrations and this PR is that the integrations are fully optional. They do not require a recompilation of BetonQuest. Instead the user updates the third party plugin in the event of a new MC release. BetonQuest is still functional (even if not all modules can be loaded) if those plugins are late to update. No action is required for a BQ maintainer.

Wolf2323 commented:
I close this PR now, as it seems to have no progress right now. If you still want this, you have multiple solutions:
1. Make an addon plugin
2. have an own Fork
3. Use `Bukkit.getServer().selectEntities()`
4. Make a compatibilety of it, so the use has to install the lib manually, instead of shading it into BQ

anyway, we only accept this PR, if a recompiling of BQ is not necessary for a new MC version.
